### PR TITLE
add `go to command` for terminal accessible buffer

### LINKED ISSUE #174079: Add symbol provider for terminal accessible buffer
Right now, we present only the command, the exit code if non-zero, and the output. 

Some questions:

- is presenting `...exited with code #` useful?
- should we add the `cwd`
- should we include the prompt

cc @jooyoungseo for any thoughts

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #174079

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/common/terminal.ts [hunk 1]
```diff
@@ -489,6 +489,7 @@ export const enum TerminalCommandId {
 	OpenWebLink = 'workbench.action.terminal.openUrlLink',
 	RunRecentCommand = 'workbench.action.terminal.runRecentCommand',
 	FocusAccessibleBuffer = 'workbench.action.terminal.focusAccessibleBuffer',
+	GoToSymbol = 'workbench.action.terminal.goToSymbol',
```
[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/common/terminal.ts:492]: Yeah I wondered about that, but that's also what we call the one in the editor
Screenshot 2023-03-14 at 11 49 54 AM

[the remark this answers is not in the document]
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/common/terminal.ts:492]: Symbol is the typical word for this in programming languages though, it's not used to describe terminal commands

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 2]
```diff
@@ -159,7 +211,6 @@ export class AccessibleBufferWidget extends DisposableStore {
 		if (existing && !existing.isDisposed()) {
 			return existing;
 		}
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:213]:
```suggestion
		}

```

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts [hunk 3]
```diff
@@ -115,7 +119,38 @@ registerAction2(class extends Action2 {
...
+		super({
+			id: TerminalCommandId.GoToCommand,
+			title: { value: localize('workbench.action.terminal.goToCommand', 'Go to Command'), original: 'Go to Command' },
+			f1: true,
+			category,
+			precondition: ContextKeyExpr.or(TerminalContextKeys.processSupported, TerminalContextKeys.terminalHasBeenCreated),
+			keybinding: [
+				{
+					primary: KeyMod.CtrlCmd | KeyMod.Shift | KeyCode.KeyO,
+					weight: KeybindingWeight.WorkbenchContrib + 2,
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:137]: this feels really nice - when in accessibility mode, if the terminal is focused and I use cmd+shift+o, it focuses the accessible buffer and opens this go to command quick pick. 

regardless of accessibility mode, now when the accessible buffer is focused, this keybinding will go to command as well

Comment (Tyriar) [reply src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:137]: cmd+shift+o was established as the keybinding for opening the detected links though, if you do this then links are no longer easily accessible?

**Comment by Tyriar** [comment]
> for restored terminals, the marker is one line off per restoration of what it should be. 

I'm not aware of this, sounds like a separate issue though. I'd expect it to reproduce when using cmd+up/down to navigate between commands?

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts [hunk 4]
```diff
@@ -115,7 +119,38 @@ registerAction2(class extends Action2 {
...
-		AccessibleBufferContribution.get(instance)?.show();
+		await AccessibleBufferContribution.get(instance)?.show();
+	}
+});
+
+registerAction2(class extends Action2 {
+	constructor() {
+		super({
+			id: TerminalCommandId.GoToCommand,
+			title: { value: localize('workbench.action.terminal.goToCommand', 'Go to Command'), original: 'Go to Command' },
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:130]: This feels a little too close to scroll to previous/next command but they do pretty different things. How about something closer to "Navigate Accessible Buffer"?

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 5]
```diff
@@ -145,6 +165,51 @@ export class AccessibleBufferWidget extends DisposableStore {
...
 	}
 
+	async showCommandQuickPick(): Promise<void> {
+		if (!this._focusedContextKey.get()) {
+			await this.show();
+		}
+		this._inQuickPick = true;
+		const commands = this._instance.capabilities.get(TerminalCapability.CommandDetection)?.commands;
+		const quickPick = this._quickInputService.createQuickPick<IQuickPickItem>();
+		const quickPickItems: IQuickPickItem[] = [];
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:175]: nit: I'd put these closer to where they're used, so quickPick before `quickPick.onDidAccept` and quickPickItems before the `for`

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts:137]: That still prevails when not in screen reader mode. I think the discoverability of this feature is more important because it enables them to traverse the buffer

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 6]
```diff
@@ -145,6 +165,51 @@ export class AccessibleBufferWidget extends DisposableStore {
...
+			return;
+		});
+		quickPick.onDidChangeActive(() => {
+			const data = quickPick.activeItems?.[0]?.meta;
+			if (data && JSON.parse(data).exitCode) {
+				this._audioCueService.playAudioCue(AudioCue.error, true);
+			}
+		});
+		quickPick.items = quickPickItems.reverse();
+		quickPick.show();
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:210]: Little thing, we normally resolve the command promise when the action is complete, like this: https://github.com/microsoft/vscode/blob/6c771e3377125c4fa7097c6a5a3d885e38dc4f88/src/vs/workbench/contrib/terminal/browser/terminalActions.ts#L402-L407

Doing this would allow an extension to chain together commands, it's would also tie the quick pick's lifecycle to the command resolving which feels nicer than it being detached from that.

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 7]
```diff
@@ -95,18 +112,21 @@ export class AccessibleBufferWidget extends DisposableStore {
...
-			if (this._accessibleBuffer.classList.contains(Constants.Active)) {
+			if (this._focusedContextKey.get()) {
 				await this._updateEditor(true);
 			}
 		}));
-		this._updateEditor();
 		this.add(this._bufferEditor.onDidFocusEditorText(async () => {
+			if (this._inQuickPick) {
+				return;
+			}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:122]: Why is this needed? I wouldn't expect the editor to get focused when the quick pick is up, and if it was, wouldn't it hide?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts [hunk 8]
```diff
@@ -35,15 +41,22 @@ export class AccessibleBufferWidget extends DisposableStore {
 	private _editorContainer: HTMLElement;
 	private _font: ITerminalFont;
 	private _xtermElement: HTMLElement;
+	private _focusedContextKey: IContextKey<boolean>;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminalAccessibleBuffer.ts:44]:
```suggestion
	private readonly _focusedContextKey: IContextKey<boolean>;
```

**Review by Tyriar — DISMISSED** [review]

**Review by Tyriar — APPROVED** [review]
